core/src/sql/statements/access.rs

```diff
@@ -0,0 +1,615 @@
+					gr.insert("token".to_owned(), Value::from(token));
+				}
+			}
+			Grant::Record(rg) => {
+				gr.insert("rid".to_owned(), Value::from(rg.rid));
+				gr.insert("jti".to_owned(), Value::from(rg.jti));
+				if let Some(token) = rg.token {
+					gr.insert("token".to_owned(), Value::from(token));
+				}
+			}
+			Grant::Bearer(bg) => {
+				gr.insert("id".to_owned(), Value::from(bg.id.to_raw()));
+				gr.insert("key".to_owned(), Value::from(bg.key));
+			}
+		};
+		res.insert("grant".to_owned(), Value::from(gr));
+
+		res
+	}
+}
+
+#[revisioned(revision = 1)]
+#[derive(Clone, Debug, Eq, PartialEq, PartialOrd, Serialize, Deserialize, Store, Hash)]
+#[cfg_attr(feature = "arbitrary", derive(arbitrary::Arbitrary))]
+#[non_exhaustive]
+pub enum Subject {
+	Record(Id),
+	User(Ident),
+}
+
+#[revisioned(revision = 1)]
+#[derive(Clone, Debug, Eq, PartialEq, PartialOrd, Serialize, Deserialize, Store, Hash)]
+#[cfg_attr(feature = "arbitrary", derive(arbitrary::Arbitrary))]
+#[non_exhaustive]
+pub enum Grant {
+	Jwt(GrantJwt),
+	Record(GrantRecord),
+	Bearer(GrantBearer),
+}
+
+#[revisioned(revision = 1)]
+#[derive(Clone, Debug, Eq, PartialEq, PartialOrd, Serialize, Deserialize, Store, Hash)]
+#[cfg_attr(feature = "arbitrary", derive(arbitrary::Arbitrary))]
+#[non_exhaustive]
+pub struct GrantJwt {
+	pub jti: Uuid,             // JWT ID
+	pub token: Option<Strand>, // JWT. Will not be stored after being returned.
+}
+
+#[revisioned(revision = 1)]
+#[derive(Clone, Debug, Eq, PartialEq, PartialOrd, Serialize, Deserialize, Store, Hash)]
+#[cfg_attr(feature = "arbitrary", derive(arbitrary::Arbitrary))]
+#[non_exhaustive]
+pub struct GrantRecord {
+	pub rid: Uuid,             // Record ID
+	pub jti: Uuid,             // JWT ID
+	pub token: Option<Strand>, // JWT. Will not be stored after being returned.
+}
+
+#[revisioned(revision = 1)]
+#[derive(Clone, Debug, Eq, PartialEq, PartialOrd, Serialize, Deserialize, Store, Hash)]
+#[cfg_attr(feature = "arbitrary", derive(arbitrary::Arbitrary))]
+#[non_exhaustive]
+pub struct GrantBearer {
+	pub id: Ident,   // Key ID
+	pub key: Strand, // Key. Will be stored but afterwards returned redacted.
+}
+
+impl GrantBearer {
+	#[doc(hidden)]
+	pub fn new() -> Self {
+		let id = random_string(GRANT_BEARER_ID_LENGTH);
+		let secret = random_string(GRANT_BEARER_KEY_LENGTH);
+		Self {
+			id: id.clone().into(),
+			key: format!("{GRANT_BEARER_PREFIX}-{id}-{secret}").into(),
+		}
+	}
+}
+
+fn random_string(length: usize) -> String {
+	let mut rng = rand::thread_rng();
+	let string: String = (0..length)
+		.map(|_| {
+			let i = rng.gen_range(0..GRANT_BEARER_CHARACTER_POOL.len());
+			GRANT_BEARER_CHARACTER_POOL[i] as char
+		})
+		.collect();
+	string
+}
+
+impl AccessStatement {
+	/// Process this type returning a computed simple Value
+	pub(crate) async fn compute(
+		&self,
+		ctx: &Context<'_>,
+		opt: &Options,
+		_doc: Option<&CursorDoc<'_>>,
+	) -> Result<Value, Error> {
+		match self {
```

Comment:
Just a nitpick, but for readability, would it help to spread these out into separate `compute` methods on the respective nested structs? So for example `AccessStatementList` would have a dedicated `compute` method

Reply:
That is a good point. I think the code will be cleaner that way :+1:

Reply:
Done in [f12b384](https://github.com/surrealdb/surrealdb/pull/4302/commits/f12b38486c24a857140fa6f8f0d9045e55f13b96).